src/nvim/strings.c

```diff
@@ -541,3 +544,847 @@ char_u *concat_str(const char_u *restrict str1, const char_u *restrict str2)
+///
+/// @return Number of bytes excluding NUL byte that would be written to the
+///         string if str_m was greater or equal to the return value.
+int vim_vsnprintf(char *str, size_t str_m, const char *fmt, va_list ap,
+                  typval_T *const tvs)
+{
+  size_t str_l = 0;
+  bool str_avail = str_l < str_m;
+  const char *p = fmt;
+  int arg_idx = 1;
+
+  if (!p) {
+    p = "";
+  }
+  while (*p) {
+    if (*p != '%') {
+      // copy up to the next '%' or NUL without any changes
+      size_t n = (size_t)(xstrchrnul(p + 1, '%') - p);
+      if (str_avail) {
+        size_t avail = str_m - str_l;
+        memmove(str + str_l, p, MIN(n, avail));
+        str_avail = n < avail;
+      }
+      p += n;
+      assert(n <= SIZE_MAX - str_l);
+      str_l += n;
+    } else {
+      size_t min_field_width = 0, precision = 0;
+      int zero_padding = 0, precision_specified = 0, justify_left = 0;
+      int alternate_form = 0, force_sign = 0;
+
+      // if both ' ' and '+' flags appear, ' ' flag should be ignored
+      int space_for_positive = 1;
+
+      // allowed values: \0, h, l, 2 (for ll), z, L
+      char length_modifier = '\0';
+
+      // temporary buffer for simple numeric->string conversion
+# define TMP_LEN 350    // 1e308 seems reasonable as the maximum printable
+      char tmp[TMP_LEN];
+
+      // string address in case of string argument
+      const char *str_arg = NULL;
+
+      // natural field width of arg without padding and sign
+      size_t str_arg_l;
+
+      // unsigned char argument value (only defined for c conversion);
+      // standard explicitly states the char argument for the c
+      // conversion is unsigned
+      unsigned char uchar_arg;
+
+      // number of zeros to be inserted for numeric conversions as
+      // required by the precision or minimal field width
+      size_t number_of_zeros_to_pad = 0;
+
+      // index into tmp where zero padding is to be inserted
+      size_t zero_padding_insertion_ind = 0;
+
+      // current conversion specifier character
+      char fmt_spec = '\0';
+
+      p++;  // skip '%'
+
+      // parse flags
+      while (true) {
+        switch (*p) {
+          case '0': zero_padding = 1; p++; continue;
+          case '-': justify_left = 1; p++; continue;
+                    // if both '0' and '-' flags appear, '0' should be ignored
+          case '+': force_sign = 1; space_for_positive = 0; p++; continue;
+          case ' ': force_sign = 1; p++; continue;
+                    // if both ' ' and '+' flags appear, ' ' should be ignored
+          case '#': alternate_form = 1; p++; continue;
+          case '\'': p++; continue;
+          default: break;
+        }
+        break;
+      }
+
+      // parse field width
+      if (*p == '*') {
+        p++;
+        const int j = tvs ? (int)tv_nr(tvs, &arg_idx) : va_arg(ap, int);
+        if (j >= 0) {
+          min_field_width = (size_t)j;
+        } else {
+          min_field_width = (size_t)-j;
```

Comment:
This looks weird, but I remember scrutinizing it long ago. And this cast just makes it explicit now.